### DISCUSSION TITLE
ANE API development

### DISCUSSION THREAD

**Discussion by henke37** [airsdk/Adobe-Runtime-Support#1873] [General]
As a previous ANE writer, I am wondering what the plans are for further development of the ANE API. It has a few omissions that I'd like to see resolved.

First off, no function to get the context data from another `ExtensionContext` object. This is needlessly annoying when your ANE offers several objects that work together.

Second, no platform specific functionality. There's no function to get a `HWND` for a `NativeWindow` object or anything else like that.

**Comment by RayGyoe** [comment]
wait for beta 50.0……

https://github.com/airsdk/Adobe-Runtime-Support/issues/236

**Comment by henke37** [comment]
Nice to see that my specific example has already been taken care of.

**Comment by ajwfrost** [comment]
Hi

Can you please clarify:
> no function to get the context data from another ExtensionContext object

If you're talking about the ActionScript side of things, then the context data should always be available via [`ExtensionContext.actionScriptData`](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/external/ExtensionContext.html#actionScriptData), and you can access this from the C side using [`FREGetContextActionScriptData`](https://help.adobe.com/en_US/air/extensions/WS29f3309f46569b4e6387539312ab5db80a1-8000.html)

Other suggestions are welcome :-) we're looking at being able to dynamically load native extensions and libraries, and to have some more advanced access into the AIR rendering pipeline.

thanks

**Reply by henke37** [reply]
I was thinking of `FREGetContextNativeData`, not `FREGetContextActionScriptData`, but they both share the same issue here. They assume that you already have a `FREContext`. Which is not the same as having a `FREObject` corresponding to a given `ExtensionContext` object.

Back in the day I did eventually kludge together a workaround, but it was more effort than a simple built in function.

**Reply by ajwfrost** [reply]
Okay I think I see.. so if you had the `FREContext` object corresponding to a particular extension, you could use this (even from another native extension, from what I can see in the code) to get at the native data via `FREGetContextNativeData`.

So then it's just a case of how to get the `FREContext` object for an extension... looking at how all the code is structured, and a little about security, I would think the best approach is if we add a C API for getting the `FREContext` C object from an `ExtensionContext` ActionScript object. So i.e. you'd need to pass in the `ExtensionContext` that you've previously created, as an argument into a function in your other extension, and then use this (as an `FREObject`) in a call to something like `FREGetFREContextFromExtensionContext`.

It seems a bit clunky, I'd have preferred a call just from C such as `FREGetFREContextFromExtensionID` where you pass in the string, but we don't appear to have a way to get at the necessary objects (there's some one-way mapping in there, and if we start adding pointers that work the other way, we could end up with garbage collection challenges...)

Let me know what you think, i.e. if it would help being able to access the `FREContext` value by using the `FREObject` representation of another `ExtensionContext` object....

thanks

**Reply by henke37** [reply]
The way I used it is to have each ExtensionContext object represent a native object. Since I was wrapping windows threads and processes, each context was for one _specific_ thread or process. As such, I had multiple ExtensionContext objects of the same type and a solution that limits each type to only one wouldn't have worked for me.

And yes, `FREGetFREContextFromExtensionContext` is exactly what I was thinking of here.

**Reply by ajwfrost** [reply]
Interesting, I hadn't really come across any ANEs where there are multiple instances of a single extension.. cunning idea!
So a `FREGetFREContextFromExtensionContext` method should be feasible, at least for a C-type extension, we can look at the Java mechanisms later if necessary!

**Comment by teotigraphix** [comment]
> we're looking at being able to dynamically load native extensions and libraries, and to have some more advanced access into the AIR rendering pipeline.

Can you elaborate a bit more on this or is it still "thought". (Like C# marshaling?)

What would more access to the rendering pipeline give _access_ to?

**Reply by ajwfrost** [reply]
The dynamic loading would mean two things:
1. Having the ability to delay all the effort of loading/initialising the libraries right at application start-up
2. Being able to separate the deployments of AIR app vs extension DLLs so that some ActionScript can dynamically look for/load an extension and query for the interface at runtime

For the rendering pipeline it basically means that you can create a display object within an ANE and have this display within the normal display list, a bit more efficiently than trying to update some bitmap data each frame. For GPU-accelerated rendering this should mean that we have a texture created and updated by an ANE that's part of the display list. 

thanks

**Reply by teotigraphix** [reply]
Huh, this sounds interesting. Any thoughts like implementing the dynamic libraries with Metadata tags like C#?

https://docs.microsoft.com/en-us/dotnet/standard/native-interop/type-marshalling

I know it's out of my ballpark thinking here but having a way to do this in AS3 would be pretty cool for adding horsepower almost invisibly.

**Reply by ajwfrost** [reply]
My initial hope was that we could replicate things like JNI in Java, where you can directly load a library from some Java code, and then it needs to have functions that match the ones that you've defined in the Java class. Dynamically calling the functions is a step further (so a bit like Java reflection). For C# libraries - or C/C++ ones - it could be possible to do similar things but there needs to be some way for the API definition to be specified so that we know what to do.. 

Taking the example from your link, we could theoretically have some ActionScript class that could have:
```
[DllImport("somenativelibrary")]
private extern function MethodA([MarshalAs(String16Ptr)] parameter : String) : int;
```

which then indicates that we've got a function called 'MethodA' which should be exposed by a library called "somenativelibrary.dll" (or "somenativelibrary.so" on Linux, or "somenativelibrary.dylib" on Mac....) and where we need to translate the string parameter into a pointer to a string of 16-bit characters (the 'int' return type doesn't have a 'MashalAs' instruction so we would assume a 32-bit return value...)

This is all feasible .. although right now the same sort of functionality could be provided by another ANE that does this marshalling and glue code. The other drawback here (imho) is that the API is fixed at compile-time so there's no flexibility.. although I guess it would be a strange use case where you've got an application trying to load and make use of an API that isn't actually known about in advance. So maybe this is most just about ease of use and clarify of code...

To actually achieve the above, it would need a fair few changes in the ActionScript compiler as well, and we've probably got some higher priority tasks to sort out first...  But yes, lots of possibilities :-)

**Reply by teotigraphix** [reply]
Thanks fro the response!

Yeah as far as;
> although right now the same sort of functionality could be provided by another ANE that does this marshalling and glue code. 

Of course but I have an audio engine that requires no state in the ANE, so having something like this for C/C++ libraries that are already handling their state would be very nice and actually pushes the needle back on maintenance time and compiling.

You end up creating things that tunnel commands like OSC, that is what I do. So there is only one dllexport C method that wraps the main API in the C++ framework for message, the rest are just bootstrap and other direct calls into the framework API. So marshaling could open up other doors.

Anyway, just thinking out loud.

**Comment by ajwfrost** [comment]
Hi

One of the updates from the AIR 50.0 pre-release version, regarding ANEs:

> AIR-5960: Adding delayed-load capability for ANEs
> To avoid having to load and initialise everything within the first frame of execution, ANEs can now be
loaded in a delayed fashion. To set this up, add a “delayLoad="true"” attribute into the
“extensionID” tag, and then – prior to accessing any of the AS3 definitions from the ANE’s library –
use the new ExtensionContext.loadExtension() method. This method can also be used to load an
ANE that’s not listed in the application descriptor file.

We probably need to get some examples up for this (and first off we're looking to update the AS3 API documentation!) - let me know of any questions.

thanks

**Reply by RayGyoe** [reply]
> Hi
> 
> One of the updates from the AIR 50.0 pre-release version, regarding ANEs:
> 
> > AIR-5960: Adding delayed-load capability for ANEs
> > To avoid having to load and initialise everything within the first frame of execution, ANEs can now be
> > loaded in a delayed fashion. To set this up, add a “delayLoad="true"” attribute into the
> > “extensionID” tag, and then – prior to accessing any of the AS3 definitions from the ANE’s library –
> > use the new ExtensionContext.loadExtension() method. This method can also be used to load an
> > ANE that’s not listed in the application descriptor file.
> 
> We probably need to get some examples up for this (and first off we're looking to update the AS3 API documentation!) - let me know of any questions.
> 
> thanks

No new content in FlashRuntimeExtensions.h file comparison?

![image](https://user-images.githubusercontent.com/336440/190467891-13984dcc-58ba-413a-914b-0762fc6e498d.png)

**Reply by ajwfrost** [reply]
Oh ... thanks, looks like we didn't copy the file from the main source code into the SDK "include" folder....! Sorry
Adding it here in a zip..
[FlashRuntimeExtensions.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/9577666/FlashRuntimeExtensions.zip)

**Reply by RayGyoe** [reply]
> 

hi @ajwfrost 
replace FlashRuntimeExtensions. h and FlashRuntimeExtensions Lib compilation error. If FREAcquireNativeWindowHandle is not called, it can be successfully compiled~

![image](https://user-images.githubusercontent.com/336440/190912966-5d6d270b-b4ee-4b7e-8494-5942836fe2c2.png)

**Reply by ajwfrost** [reply]
Thanks for checking this, yes we had some exports missing in the library files, correcting that for our next version..
